### ISSUE TITLE
How to change the color of all links in one PDF

### ISSUE THREAD

**Issue by jiezhoujenny** [pymupdf/PyMuPDF#550]
I have a PDF file that I need to change all the links' color from black to blue.
Tried link.setColors, seems doesn't work, can you please provide some sample code on this?
I have code like:

import fitz

red = (1, 0, 0)
blue = (0, 0, 1)
gold = (1, 1, 0)
green = (0, 1, 0)

def setLinkColors():
    doc = fitz.open("test.pdf")
    page = doc[0]  
    ln = page.firstLink
................ how to change color here?

Many thanks

**Comment by JorjMcKie** [comment]
Should be something like this:

```python
for page in doc:
    for link in page.links():
        link.setColors(stroke=...)  # fill color not supported for links.
        link.setBorder(...) # also set border width, because it is 0 by default
```
That's it.
Please note, that for links all you can achieve is setting border line properties. You cannot change the text or background color.
There also is no ``update`` method for links so the above code directly applies the changes.

**Comment by JorjMcKie** [comment]
I suspect you meant to change the **_text color_** within a link.
This is **_not possible_** - I may have been not explicit enough with my previous post. The text color is not part of the link object definition, but it is specified as part of the page's text paintings. So (in theory) you would have to go there and make the change.
And there is no (direct) way to do this in PyMuPDF.

**Comment by jiezhoujenny** [comment]
> I suspect you meant to change the **_text color_** within a link.
> This is **_not possible_** - I may have been not explicit enough with my previous post. The text color is not part of the link object definition, but it is specified as part of the page's text paintings. So (in theory) you would have to go there and make the change.
> And there is no (direct) way to do this in PyMuPDF.

Thank you so much for your reply. Yes, what I meant is to change the text color within a link.
As you suggested, I might able to change the text color via text painting, but is there any way to extract the link text? I don't see the link class has any attribute to get the text of link.

**Comment by JorjMcKie** [comment]
As I said: the connection between a link object on the one hand and what you see on the page inside the "hot area" of the link (i.e. ``link.rect`` or ``link['from']``) is only loosely coupled.
This has the following reasons:
* A link rectangle can cover some text, which may be something like "www.something.xyz" or just **_any text_** else.
* A link rectangle can cover an image - you will find no text at all inside that rectangle.

You can of course use one of the text extraction methods of PyMuPDF and extract the text ["under"](https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/examples/textboxtract.py) the link rectangle.
Once this is done, you will have to parse the page's `/Contents`. This is written in a special, PDF-specific language, explained in the PDF manual on page 985 (Appendix A Operator Summary).
You can access the contents via `page.readContents()`. This a `bytes` object. Text objects within this source are wrapped by keywords `b"BT"´ (begin text) and `b("ET")` (end text). Inside these markers you should find the text and its color definition.
I must warn you: this is a very complex, cumbersome undertaking!

If you happen to know the exact color definiton of the link texts of your file, there might be a shortcut.
* Take an example page with some link
* execute `page.cleanContents()`.
* execute `cont=page.readContents()` and save the result in a text file:
```page
out = open("cont.txt", "wb")
out.write(cont)
out.close()
```
* open the created file and try to find one of the link texts in it. You should see a line nearby containing `a b c RG`, where a, b, c are some floating point numbers, each between 0 and 1 (inclusive)
* If you can find such a line, then (a, b, c) is the RGB color used for links. In the hope, that this specific color is never used for anything else, one can **_replace every such line_** by the line `0 0 0 RG` (which stands for RGB black).
* Once this is done, the modified contetns can be written back to the PDF.

... following up with some example (hopefully)

**Comment by jiezhoujenny** [comment]
Many thanks!! I will give a try as you commented, appreciate!

**Comment by JorjMcKie** [comment]
Just another comment:
Understanding your intention and considering the extremely difficult approach we discussed above, you may want to also think about alternatives, which are very much easier to achieve:

1. You can add text highlighting annotations "over" each link: use the link rectangle and do `page.addHighlightAnnot(rect)`. This will let the link text look like being highlighted with a text marker. Or instead underline the text via `page.addUnderlineAnnot(rect)`. Annotations can also later be removed - they are not "permanent".
2. You can also achieve a permanent effect like with annotations: `page.drawRect(rect, fill=(1,1,0), overlay=False)` will put a permanent yellow rectangle underneath the link text. Similarly, `page.drawLine(rect.bl, rect.br, color=(0,0,1))` will underline with a permanent blue line.

```python
>>> page=doc[...]
>>> shape=page.newShape()
>>> for link in page.links():
	rect = link["from"]
	shape.drawLine(rect.bl, rect.br)

Point(246.79299926757812, 438.364013671875)
Point(250.7779998779297, 438.364013671875)
Point(540.9959716796875, 575.8489990234375)
Point(101.77799987792969, 587.8040161132812)
Point(105.76300048828125, 587.8040161132812)
>>> shape.finish(color=(0,0,1))  # blue
>>> shape.commit()  # lines in foreground are no problem
>>> doc.save(...)
```

will add permanent underlines like this:

![grafik](https://user-images.githubusercontent.com/8290722/86781513-00a30880-c02c-11ea-852d-25352d31c989.png)
--------------------
Similarly with permanent background (yellow in this case):
```python
>>> for link in page.links():
	rect = link["from"]
	shape.drawRect(rect)

	
Point(212.6909942626953, 425.8710021972656)
Point(244.80099487304688, 425.8710021972656)
Point(512.9920043945312, 564.9450073242188)
Point(71.00399780273438, 575.31103515625)
Point(99.78500366210938, 575.31103515625)
>>> shape.finish(fill=(1,1,0))  # make rectangles yellow
>>> shape.commit(overlay=False)  # put them all in background
>>> doc.save(...)
```
Will give this appearance:

![grafik](https://user-images.githubusercontent.com/8290722/86781941-87f07c00-c02c-11ea-829b-9d219f3f3489.png)


This way you can achieve your goal (make links with black text more visible) with a minimum of effort.

**Comment by jiezhoujenny** [comment]
It helped me a lot,  I will try as you advised, thank you for your great work.
